src/RageTimer.cpp

```diff
@@ -51,14 +51,21 @@ static std::uint64_t GetTime( bool /* bAccurate */ )
 #endif
 }
 
-float RageTimer::GetTimeSinceStart( bool bAccurate )
+/* In the past, this was float instead of double to maintain compatibility
+with legacy hardware that can't use the double type. On modern hardware,
+performance of double is good or better than mixing 32 bit and 64 bit wide
+types, and maintaining precision is directly related to the stability of clock
+sync during gameplay. Too much error here will manifest as a drastic shift
+in the game's sync (enough to push a Fantastic into the Great window or worse).
+Incorrect math here will manifest as a consistent sync offset in game, similar
+to the effect of changing the global offset. */
+double RageTimer::GetTimeSinceStart(bool bAccurate)
 {
-	std::uint64_t usecs = GetTime( bAccurate );
+	std::uint64_t usecs = GetTime(bAccurate);
 	usecs -= g_iStartTime;
-	/* Avoid using doubles for hardware that doesn't support them.
-	 * This is writing usecs = high*2^32 + low and doing
+	/* This is writing usecs = high*2^32 + low and doing
 	 * usecs/10^6 = high * (2^32/10^6) + low/10^6. */
-	return std::uint32_t(usecs>>32) * 4294.967296f + std::uint32_t(usecs)/1000000.f;
+	return (usecs >> 32) * 4294.967296 + (usecs) / 1000000.0;
```

Comment:
If I understand this function correctly, it's just converting GetTime's usecs to seconds?

Can we remove the obscure bitshift / float maths here?

As a very quick check, throwing this into compiler explorer (assuming I got the maths right), shows that this version is actually more instructions (58 lines vs 35 lines) than using doubles and letting the compiler do its thing (at least for the default/gcc settings used here. it will vary across compilers).

Any thoughts? Is that worth a further check for performance/similar? Could be a completely separate issue for later, since as long as this version still produces the same result it won't actually impact the sync fix.

Just struck me as a little odd - Anything where we're mixing bitshifts and magic float numbers together seems like trying to beat the compiler, and that rarely results in the fastest code.

![image](https://github.com/itgmania/itgmania/assets/11091460/17129bba-3653-4108-a9ab-5a55344f05b9)

Reply:
Looking at the StepMania 3.95 code (which was arcade ITG), we actually aren't doing any of this weird bitshift math either: https://github.com/sigatrev/Stepmania-3.95/blob/master/RageTimer.cpp#L45

Reply:
Thanks to both of you for pointing this out.

I'll wait to see how the tests of my `patch-3-chrono` branch come back before changing this, as the bitshift math is completely removed there:

https://github.com/sukibaby/itgmania/blob/patch-3-chrono/src/RageTimer.cpp

Reply:
Ok, so initial tests of `patch-3-chrono` weren't great. It seems to drift and then go back multiple times per song, and the combo counter is strangely laggy, so probably there's something else in the code messing out the time `std::chrono::steady_clock` is giving back.

For that reason, I'll be continuing work on this branch.

Since the efficiency and accuracy of this function is key to maintaining good sync and not drifting, I've implemented @geefr suggestion of dividing by 1000000. I wouldn't be surprised if the magic numbers are adding a strange amount of error here and there, too.

By the way, I haven't fully understood yet all of where the boolean `bAccurate` is being passed around to and from. Does anyone have an idea why the boolean is partly commented out in ITGMania as-is? Is it needed?

https://github.com/itgmania/itgmania/blob/0ebaffa6f174e79039267cbd600eb78f93222cdc/src/RageTimer.cpp#L38

Reply:
I can try and check later but I think accurate is always true.

For the chrono issue I vaguely recall steady clock not being monotonic/steady in reality on windows - here we need a high resolution timer, but also once that cannot jump around or drift ever...can't remember what codebase that was in though as it was a while ago

Reply:
> I can try and check later but I think accurate is always true.
> 
> For the chrono issue I vaguely recall steady clock not being monotonic/steady in reality on windows - here we need a high resolution timer, but also once that cannot jump around or drift ever...can't remember what codebase that was in though as it was a while ago

Thank you for checking on bAccurate. 🙂 

I can try with `std::chrono::high_resolution_clock` instead of `steady_clock` and see if it is better.

Reply:
> I can try and check later but I think accurate is always true.

I looked into it a little more. We haven't used that boolean in 12 years:

https://github.com/stepmania/stepmania/commit/52df8ea9818b252e92210c974cc9105c3a7955a2

I'm ok with leaving it as-is for now, though I wish I could find an explanation of why it was commented out the way it was...

Reply:
'I can try with std::chrono::high_resolution_clock instead of steady_clock and see if it is better'

To avoid any more complexity here - Feel free to tag me on the other issue / chase via discord for a chat. The issue there would then be that high_resolution_clock isn't guaranteed to be monotonic either - There might not be a good choice from chrono at all since SM needs both high resolution and monotonic, and either high_resolution or steady may not provide both.

steady_clock _should_ be monotonic but A: On some platforms it's only accurate to 10ms or so and B: On older versions of visual studio it's not monotonic, even if it claims to be. I think it was also fixed in a minor version bump to MSVC, so checking _MSC_VER_ or similar might not be enough to guarantee it works (this will need quite a bit of research - see here but it's a rabbit hole to worry about later: https://developercommunity.visualstudio.com/t/stl-steady-clock-not-steady/1105325)

When I hit this one before (A cross platform graphics engine, pretty sure it was a work thing) we just stuck to the platform-specific timers. Yes we had to #ifdef windows all over the place, but the behaviour was stable across all versions - C++ chrono is nice for general use, but might not be the best choice for engine level things..

Reply:
> 'I can try with std::chrono::high_resolution_clock instead of steady_clock and see if it is better'
> 
> To avoid any more complexity here - Feel free to tag me on the other issue / chase via discord for a chat. The issue there would then be that high_resolution_clock isn't guaranteed to be monotonic either - There might not be a good choice from chrono at all since SM needs both high resolution and monotonic, and either high_resolution or steady may not provide both.
> 
> steady_clock _should_ be monotonic but A: On some platforms it's only accurate to 10ms or so and B: On older versions of visual studio it's not monotonic, even if it claims to be. I think it was also fixed in a minor version bump to MSVC, so checking _MSC_VER_ or similar might not be enough to guarantee it works (this will need quite a bit of research - see here but it's a rabbit hole to worry about later: https://developercommunity.visualstudio.com/t/stl-steady-clock-not-steady/1105325)
> 
> When I hit this one before (A cross platform graphics engine, pretty sure it was a work thing) we just stuck to the platform-specific timers. Yes we had to #ifdef windows all over the place, but the behaviour was stable across all versions - C++ chrono is nice for general use, but might not be the best choice for engine level things..

This ended up working out very nicely on Windows along with the RageTimer improvements: https://github.com/itgmania/itgmania/pull/234

I tried `steady_clock` and `high_resolution_clock`, but neither performed nearly as well as QPC set up how the Microsoft documentation suggests to do it.